### ISSUE TITLE
How to save the prediction result on training set and test set?

### ISSUE THREAD

**Issue by Pachark** [zezhishao/BasicTS#44]
Great job! But I didn't find how to save the specific prediction results in config.

**Comment by zezhishao** [comment]
Thanks for your attention.
You can find them in the training log files.
Specifically, all training logs will be saved in `checkpoints/MODEL_EPOCH/MD5_of_config_file`. 
For example, `checkpoints/D2STGNN_100/85f8fd856810f2a2ce63b897f141864f`.

Here is the training log of D2STGNN on the METR-LA dataset:

```text
2022-09-02 15:01:54,809 - easytorch-training - INFO - Result <train>: [train_time: 162.60 (s), lr: 1.56e-05, train_MAE: 2.5934, train_RMSE: 5.1506, train_MAPE: 0.0671]
2022-09-02 15:01:54,811 - easytorch-training - INFO - Start validation.
2022-09-02 15:02:03,768 - easytorch-training - INFO - Result <val>: [val_time: 8.96 (s), val_MAE: 2.6475, val_RMSE: 5.1250, val_MAPE: 0.0718]
2022-09-02 15:02:03,836 - easytorch-training - INFO - Checkpoint checkpoints/D2STGNN_100/85f8fd856810f2a2ce63b897f141864f/D2STGNN_best_val_MAE.pt saved
2022-09-02 15:02:21,336 - easytorch-training - INFO - Evaluate best model on test data for horizon 1, Test MAE: 2.0811, Test RMSE: 3.6146, Test MAPE: 0.0490
2022-09-02 15:02:21,339 - easytorch-training - INFO - Evaluate best model on test data for horizon 2, Test MAE: 2.3710, Test RMSE: 4.3777, Test MAPE: 0.0582
2022-09-02 15:02:21,341 - easytorch-training - INFO - Evaluate best model on test data for horizon 3, Test MAE: 2.5543, Test RMSE: 4.8949, Test MAPE: 0.0650
2022-09-02 15:02:21,343 - easytorch-training - INFO - Evaluate best model on test data for horizon 4, Test MAE: 2.6967, Test RMSE: 5.3210, Test MAPE: 0.0705
2022-09-02 15:02:21,345 - easytorch-training - INFO - Evaluate best model on test data for horizon 5, Test MAE: 2.8047, Test RMSE: 5.6422, Test MAPE: 0.0748
2022-09-02 15:02:21,348 - easytorch-training - INFO - Evaluate best model on test data for horizon 6, Test MAE: 2.9020, Test RMSE: 5.9155, Test MAPE: 0.0790
2022-09-02 15:02:21,349 - easytorch-training - INFO - Evaluate best model on test data for horizon 7, Test MAE: 2.9928, Test RMSE: 6.1642, Test MAPE: 0.0827
2022-09-02 15:02:21,351 - easytorch-training - INFO - Evaluate best model on test data for horizon 8, Test MAE: 3.0662, Test RMSE: 6.3694, Test MAPE: 0.0856
2022-09-02 15:02:21,352 - easytorch-training - INFO - Evaluate best model on test data for horizon 9, Test MAE: 3.1361, Test RMSE: 6.5526, Test MAPE: 0.0885
2022-09-02 15:02:21,354 - easytorch-training - INFO - Evaluate best model on test data for horizon 10, Test MAE: 3.2093, Test RMSE: 6.7209, Test MAPE: 0.0913
2022-09-02 15:02:21,355 - easytorch-training - INFO - Evaluate best model on test data for horizon 11, Test MAE: 3.2741, Test RMSE: 6.8814, Test MAPE: 0.0939
2022-09-02 15:02:21,357 - easytorch-training - INFO - Evaluate best model on test data for horizon 12, Test MAE: 3.3431, Test RMSE: 7.0353, Test MAPE: 0.0965
2022-09-02 15:02:21,368 - easytorch-training - INFO - Result <test>: [test_time: -17.53 (s), test_MAE: 2.8693, test_RMSE: 5.8797, test_MAPE: 0.0779]
2022-09-02 15:02:21,410 - easytorch-training - INFO - Checkpoint checkpoints/D2STGNN_100/85f8fd856810f2a2ce63b897f141864f/D2STGNN_078.pt saved
2022-09-02 15:02:21,410 - easytorch-training - INFO - The estimated training finish time is 2022-09-02 16:11:23
```

**Comment by Pachark** [comment]
Maybe I didn't clarify the question clearly. I wonder how to save the predicted value of all sample points so I can visualize the results. I just found out that I should modify `runners.base_tsf_runner.py` to achieve this function, right?

**Comment by zezhishao** [comment]
Yes, you are right!

You can save the prediction values of the **test set** by modifying [`test` function](https://github.com/zezhishao/BasicTS/blob/f776bcd3da3441d80a43ef8a4dae68981f58a0fb/basicts/runners/base_tsf_runner.py#L277) in `runners.base_tsf_runner.py`.

However, saving the results of the training set might be a bit complicated. You need to modify the [`train` function](https://github.com/cnstark/easytorch/blob/1ec3056f0609551d4fb433185bb4c44e4d257fce/easytorch/core/runner.py#L297) of easytorch.

**Comment by Pachark** [comment]
This is great! 

Thank you for your quick reply @zezhishao  ☺️

**Comment by zezhishao** [comment]
You're welcome! 
If you have any questions, please feel free to open a new issue.
